### DISCUSSION TITLE
Is it possible to set up a winkModel (using the `wink-eng-lite-web-model`) in a Worker?

### DISCUSSION THREAD

**Discussion by SkepticMystic** [winkjs/wink-nlp#68] [Help on coding] [answered]
I'm trying to set up a web worker to handle some nlp stuff.
The project I'm working on requires me to use the `web-model`.

I've tried a few different approaches, but I always end up getting this error:
![image](https://user-images.githubusercontent.com/70717676/146889977-48316c8a-fd79-4e94-8dac-4404aa6b1751.png)

Pointing to this section of code:

![](https://media.discordapp.net/attachments/799660344627953725/922751804402450452/unknown.png)

I've tried importing wink in different ways, but I don't think that's the issue. I've also configured esbuild to bundle it for `platform: "browser"`, but that hasn't changed the error.
I'm not sure if this issue is related to wink, but any help would be appreciated.

This worker code should reproduce the error:

```ts
self.onmessage = async function (e) {
	const model = require("wink-eng-lite-web-model");
	const wink = require("wink-nlp");
	const winkModel = wink(model);

	self.postMessage(e.data);
};
```

**Comment by sanjayaksaxena** [comment]
Have you considered requiring the package outside the function?

**Comment by SkepticMystic** [comment]
Yes, importing at the top of the script gives the same error

**Comment by sanjayaksaxena** [comment]
See if this gives you any clue: https://observablehq.com/@winkjs/how-to-run-wink-nlp-on-observable?collection=@winkjs/winknlp-recipes

**Reply by sanjayaksaxena** [reply]
Apart from the above, we had tested it on browser using Browserify — https://dev.to/winkjs/how-to-run-nlp-in-the-browser-2ca6

**Reply by SkepticMystic** [reply]
I already do what the first link suggests.
For the second link, I browserified the worker, but then the output file no longer exports the worker. I added `export {self: myWorker}`, but then Browserify won't run, saying I need `sourceType: 'module'`

**Reply by sanjayaksaxena** [reply]
We have tried following and it works:

### index.html
```html
<html lang="en">
  <head>
    <title>Web Workers Demo</title>
    <meta charset="UTF-8" />
  </head>

  <body>
    <div id="app"></div>
    <h1>Web Workers Demo — Basic NLP</h1>

    <script src="main.js"></script>
  </body>
</html>
```

### main.js
```javascript
const worker = new Worker("worker.js");

worker.onmessage = e => {
  const message = e.data;
  console.log(`[From Worker (Tokens)]: ${message}`);
};

worker.postMessage("Hello World!");
```

### worker.js
```javascript
onmessage = async function ( e ) {
  const winkNLP = (await import('https://cdn.skypack.dev/wink-nlp')).default;
  const model = (await import('https://cdn.skypack.dev/wink-eng-lite-web-model')).default;
  const nlp = winkNLP(model);

  const message = nlp.readDoc(e.data).tokens().out();
  console.log(`[From Main (Text)]: ${e.data}`);

  postMessage(JSON.stringify(message));
};
```

Here is the screenshot of output on Safari browser:
<img width="584" alt="image" src="https://user-images.githubusercontent.com/29990/146933004-ec5f7ecb-2aa1-4e3b-9f9b-83d66b596256.png">

_Have used Observable notebook styled imports — as mentioned earlier_. Safari version is **15.1** running on Macbook Pro with OS version **12.0.1**.

**Reply by SkepticMystic** [reply]
This works!
Ideally, I wouldn't need to import it from the web everytime. I'm building a plugin for a mostly-offline app, but this will mean users need an internet connection.
But this is a fine compromise

**Reply by sanjayaksaxena** [reply]
You can use `import` `from` instead of `await import` inside a `async function`, then it should work. We followed the structure given in your original question.

**Reply by SkepticMystic** [reply]
But I would still need to fetch it from a remote.
I think we have found the issue. The local `wink-eng-lite-web-model` uses `Buffer`, which is causing the error.
But when I import it from the remote, as in your last answer, it works. So the remote version must not be using Buffer?

**Reply by SkepticMystic** [reply]
Can't this module use an `ArrayBuffer` instead? To be non-Node compatible
![image](https://user-images.githubusercontent.com/70717676/146960427-c6c63673-0ae8-48ea-81a8-dcecacbfecae.png)

**Reply by SkepticMystic** [reply]
That might not be the cause of the issue. But basically, if the web-model fetched from the skypack remote works, but the local version doesn't, then something must be different

**Comment by sanjayaksaxena** [comment, marked as answer]
In that case simply browserify your worker and it should work:

### nlp-main.js
```javascript
const winkNLP = require( 'wink-nlp' );
const model = require( 'wink-eng-lite-web-model' );
const nlp = winkNLP( model )

onmessage = function ( e ) {
  const message = nlp.readDoc(e.data).tokens().out();
  console.log(`[From Main (Text)]: ${e.data}`);

  postMessage(JSON.stringify(message));
};
```

The above file should be browserify using following command:

```sh
browserify nlp-main.js -o nlp.js
```

**Reply by SkepticMystic** [reply]
Haha then I get this problem:
> I browserified the worker, but then the output file no longer exports the worker. I added `export {self: myWorker}`, but then Browserify won't run, saying I need `sourceType: 'module'`

**Reply by sanjayaksaxena** [reply]
Please see above updated reply, there is no export.

**Reply by sanjayaksaxena** [reply]
Giving the entire code again for you:

### index.html
```html
<html lang="en">
  <head>
    <title>Web Workers Demo</title>
    <meta charset="UTF-8" />
  </head>

  <body>
    <div id="app"></div>
    <h1>Web Workers Demo with Browserify — Basic NLP</h1>

    <script src="main.js"></script>
  </body>
</html>
```

### nlp-main.js
```javascript
const winkNLP = require( 'wink-nlp' );
const model = require( 'wink-eng-lite-web-model' );
const nlp = winkNLP( model )

onmessage = function ( e ) {
  const message = nlp.readDoc(e.data).tokens().out();
  console.log(`[From Main (Text)]: ${e.data}`);

  postMessage(JSON.stringify(message));
};
```

### main.js
```javascript
const worker = new Worker("nlp.js" );

worker.onmessage = e => {
  const message = e.data;
  console.log(`[From Worker (Tokens)]: ${message}`);
};

worker.postMessage("Hello World!");

worker.postMessage("Hi how are you!");
```

Browserify using the following command:
```sh
browserify nlp-main.js -o nlp.js
```

Here is the screenshot:
<img width="817" alt="image" src="https://user-images.githubusercontent.com/29990/146968246-55f619a3-4b0d-4f16-8629-2d54fb1d5f9e.png">

**Comment by SkepticMystic** [comment]
I see, I see.
So I browserified the worker. The issue now is that the app I am building the plugin for is trying to resolve the worker script path as a URI from its perspective. This is obviously not an issue on your side, so I will speak to the dev to see what can be done.

**Reply by sanjayaksaxena** [reply]
all the best 👍

**Reply by SkepticMystic** [reply]
I really appreciate all your help, Sanjay :)

**Reply by sanjayaksaxena** [reply]
That is the beauty & power of open source community. Please mark the suggested solution as answer to make it useful to our community.

**Reply by SkepticMystic** [reply]
I have marked it as the answer.
I do still think that it's strange that the remote web-model behaves differently to the local npm-installed one. Should the code not be the same?

**Reply by SkepticMystic** [reply]
I tried to get the raw js from the remote skypack, but this url doesn't show the code: https://cdn.skypack.dev/pin/wink-eng-lite-web-model@v1.3.2-AncWcCKowguVHDW4e1Md/mode=imports,min/optimized/wink-eng-lite-web-model.js

Do you know if I can just copy the remote code into my local file?

**Reply by sanjayaksaxena** [reply]
Why not use the **`browserified worker`**? This file can be stored locally on your device.

**Reply by SkepticMystic** [reply]
There are lots of issues in trying to use the browserified worker in the context of the app I'm working in.
If I could just use the remote skypack code locally, I know that will work, and my users don't have to fetch it everytime

**Reply by sanjayaksaxena** [reply]
I guess best is to lookup skypack documentation or ask them...

**Comment by sanjayaksaxena** [comment]
Hi @SkepticMystic 

We have release a new version of our `wink-eng-lite-web-model` (Version 1.3.3 on NPM), where calls to Node.js' `Buffer.from()` API have been completely removed. 

Hope it helps you further.

**Reply by SkepticMystic** [reply]
Thanks for the ping @sanjayaksaxena :D
I'll be sure to try it out when I can!
